### ISSUE TITLE
PyTorch export crash with onnx >= 1.8.0 unless import onnx first on windows

### ISSUE THREAD

**Issue by BowenBao** [onnx/onnx#3493]
Similar to #2808

The script crashes when onnx is imported after pytorch. The script completes successfully when onnx is imported before pytorch, or when onnx installed version <= 1.7.0

```python
# import onnx  # uncomment and pass
import torch
from onnx import ModelProto

class M(torch.nn.Module):
    def forward(self, x, y):
        return x + y

x = torch.randn(2, 3)
y = torch.randn(2, 3)

import io
f = io.BytesIO()
torch.onnx.export(M(), (x, y), f, verbose=True, input_names=['x', 'y'])
```

Both packages are installed like below
```
pip install torch==1.8.1+cpu -f https://download.pytorch.org/whl/torch_stable.html
pip install onnx==1.8
```

Based on previous issues, I suspect this is also related to different protobuf versions. Can someone explain what is going on underneath, and how to resolve & prevent this in the future?

**Comment by jcwchen** [comment]
Hi @BowenBao,
On Mac I solved a similar issue by upgrading protobuf-compiler (to 3.11.3) and MacOS (10.15). However, on Windows, basically ONNX==1.7 and ONNX>=1.8 are using the same way to install protobuf (3.11.3 static protobuf) as dependency so it looks weird to me that why only ONNX 1.7 works... I have questions:
1. Did you only see this crash on Windows?
2. Will you see this crash if you pip install torch then install ONNX from source (current main branch)?

Thank you!

**Comment by jcwchen** [comment]
BTW, I can repro it with ONNX 1.9.0 as well..

**Comment by BowenBao** [comment]
hi @jcwchen, 
1. yes only on windows.
2. yes observe same behavior with ONNX built from source, with libprotobuf & protobuf installed by `conda install -c conda-forge  libprotobuf=3.11.3 protobuf`.

**Comment by askhade** [comment]
@BowenBao : how are you installing onnx? from pypi or conda?

**Comment by BowenBao** [comment]
> 
> 
> @BowenBao : how are you installing onnx? from pypi or conda?

pypi via `pip install onnx==1.8`

**Comment by jcwchen** [comment]
Hi @BowenBao,
When trying to figure out whether this issue still exists in upcoming ONNX 1.10.0, I have some new findings:

1. If using static protobuf on Windows, any ONNX version+PyTorch with this script won't bump into any issue. My previous experiment was done in a conda environment (using shared Protobuf). In conda, it still only works with ONNX 1.7, and other versions (even the upcoming 1.10.0) will still crash.
2. I found there is still a slight difference between ONNX 1.7 and other versions about building Protobuf for Windows wheel:
    - ONNX 1.7 build static latest Protobuf from vcpkg (It's quite hard to trace which version was the latest vcpkg package at that time... should be 3.10.X or 3.11.X)
    - ONNX 1.8.0, 1.8.1, 1.9.0 build static 3.11.3 protobuf from source
    - ONNX 1.10.0 build static 3.16.0 protobuf from source
 
Since the used protobuf version is different for ONNX 1.7, it can explain that why only ONNX 1.7 can work in conda environment (using shared protobuf).

In short, using static protobuf should resolve this crash. And, we do recommend using static protobuf with ONNX, especially on Windows. Otherwise, there are potential issues about using different Protobuf versions as shared library... @BowenBao Do you know which Protobuf version that Pytorch is using? Thank you.

**Comment by garymm** [comment]
PyTorch's ProtoBuf library is being upgraded to 3.13.1 in https://github.com/pytorch/pytorch/pull/62571

**Comment by malfet** [comment]
The reason why we have specific version of ProtoBuf associated with PyTorch, is that we compile all protobuf dependencies with local visibility, so it should not conflict with other versions.
I wonder if what is happening here is onnx symbol conflicts(i.e. same symbol is being exported from libtorch_cpu as well as from onnx runtime). Next time it happens, can someone please share a crashdump?

**Comment by jcwchen** [comment]
Thank you both for the updates! 

More updates from my experiments: 
- I ran the same release CI with the same code base as ONNX 1.7 and got the similar wheel. (The only difference is the latest protobuf version from vcpkg is different when releasing ONNX 1.7) However, the produced ONNX 1.7 wheel will crash with PyTorch as well. I tried several different versions of protobuf, but none of them could work.
- After more experiments, I found that actually using local **static** protobuf will bump into this crash as well, sorry that I need to correct my previous statement
- Only ONNX <= 1.7 from PyPI won't encounter this crash.


>  is that we compile all protobuf dependencies with local visibility, so it should not conflict with other versions.

Could you please confirm that is it true now? Because I saw there is an old issue about it: https://github.com/pytorch/pytorch/issues/14573 


> Next time it happens, can someone please share a crashdump?

I am still able to repro it by using the script provided by Bowen, but it crashed without any error message. I used a traceback tool and found it crashed at this line: https://github.com/pytorch/pytorch/blob/37c1f4a7fef115d719104e871d0cf39434aa9d56/torch/nn/modules/module.py#L860  @malfet Please let me know if there is something I can do from my side to help you identify this issue. Thank you!

**Comment by snnn** [comment]
I can't repro it. I didn't use conda. I use python 3.9 from python.org.
```
C:\a>pip install torch==1.8.1+cpu -f https://download.pytorch.org/whl/torch_stable.html
Looking in links: https://download.pytorch.org/whl/torch_stable.html
Collecting torch==1.8.1+cpu
  Downloading https://download.pytorch.org/whl/cpu/torch-1.8.1%2Bcpu-cp39-cp39-win_amd64.whl (190.4 MB)
     |████████████████████████████████| 190.4 MB 29 kB/s
Requirement already satisfied: typing-extensions in c:\python39\lib\site-packages (from torch==1.8.1+cpu) (3.7.4.3)
Requirement already satisfied: numpy in c:\python39\lib\site-packages (from torch==1.8.1+cpu) (1.20.2)
Installing collected packages: torch
Successfully installed torch-1.8.1+cpu
WARNING: You are using pip version 20.2.3; however, version 21.2.2 is available.
You should consider upgrading via the 'c:\python39\python.exe -m pip install --upgrade pip' command.

C:\a>pip install onnx
Collecting onnx
  Downloading onnx-1.10.1-cp39-cp39-win_amd64.whl (11.4 MB)
     |████████████████████████████████| 11.4 MB 6.8 MB/s
Requirement already satisfied: typing-extensions>=3.6.2.1 in c:\python39\lib\site-packages (from onnx) (3.7.4.3)
Requirement already satisfied: numpy>=1.16.6 in c:\python39\lib\site-packages (from onnx) (1.20.2)
Requirement already satisfied: protobuf in c:\python39\lib\site-packages (from onnx) (3.15.7)
Requirement already satisfied: six in c:\python39\lib\site-packages (from onnx) (1.15.0)
Installing collected packages: onnx
Successfully installed onnx-1.10.1
```

**Comment by jcwchen** [comment]
@snnn Did you run the Python script as follows with ONNX>=1.8.0 and PyTorch==1.8.1:
```
import torch
from onnx import ModelProto

class M(torch.nn.Module):
    def forward(self, x, y):
        return x + y

x = torch.randn(2, 3)
y = torch.randn(2, 3)

import io
f = io.BytesIO()
torch.onnx.export(M(), (x, y), f, verbose=True, input_names=['x', 'y'])
```
And it worked without crash?  Which protobuf version do you use?  Thank you!

**Comment by snnn** [comment]
![image](https://user-images.githubusercontent.com/856316/128447869-755bdc05-0a9b-4ab3-aecc-1df732c59e83.png)
No crash at all.

> Which protobuf version do you use?

None. It doesn't apply to me. I got the two python packages from pypi. I didn't build any. So I don't need to install protobuf.

**Comment by snnn** [comment]
Now I can guess, it must because someone dynamically linked to protobuf and there is a version conflict.

**Comment by malfet** [comment]
Protobuf update inside pytorch is blocked as it causes crash every time ONNX is used, i.e. when calling something as simple as:
```
python -c "import torch;import io;torch.onnx.export(torch.nn.Linear(3,3), (torch.rand(3, 3), ), io.BytesIO(), verbose=True)"
```
@garymm is there an environment variable once can use to  additional debug output while onnx export is running?

**Comment by askhade** [comment]
> Now I can guess, it must because someone dynamically linked to protobuf and there is a version conflict.

Yes that must be it.

**Comment by garymm** [comment]
@malfet I would try running it under a debugger (gdb, lldb).

Something like `gdb python`, then `run  -c "import torch;import io;torch.onnx.export(torch.nn.Linear(3,3), (torch.rand(3, 3), ), io.BytesIO(), verbose=True)"`

If that doesn't work, you could try setting `PYTORCH_JIT_LOG_LEVEL` to enable logging on all the files under the [onnx dir](https://github.com/pytorch/pytorch/tree/3979cb0656fe2d8b0445768a769bd624b10778b5/torch/csrc/jit/passes/onnx)
https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/OVERVIEW.md#jit-logging

**Comment by malfet** [comment]
@garymm unfortunately crash is windows only, where code is by default missing any symbolic information and neither gdb nor lldb work...

**Comment by jcwchen** [comment]
Update: I found if ONNX is built with Debug mode, torch.onnx.export won't bump into this crash (even in a conda environment): [ONNX wheel from main branch with debug mode](https://github.com/jcwchen/onnx/suites/3444194129/artifacts/81688585).

The crash log is ([full log](https://github.com/onnx/onnx/files/6952072/crash_log.txt))
```
53d4:4228 @ 194465281 - LdrpGetProcedureAddress - INFO: Locating procedure "PyInit__message" by name
(53d4.4228): C++ EH exception - code e06d7363 (first chance)
53d4:4228 @ 194465328 - LdrpGetProcedureAddress - INFO: Locating procedure "FlsGetValue" by name
(53d4.4228): Access violation - code c0000005 (first chance)
First chance exceptions are reported before any exception handling.
This exception may be expected and handled.
*** WARNING: Unable to verify checksum for C:\Users\chunwche\Anaconda3\lib\site-packages\onnx\onnx_cpp2py_export.cp37-win_amd64.pyd
onnx_cpp2py_export_cp37_win_amd64!PyInit_onnx_cpp2py_export+0x130cfe:
00007ffe`7da2d2fe 83780c00        cmp     dword ptr [rax+0Ch],0 ds:00fffdbc`60578e0c=????????

```

**Comment by jcwchen** [comment]
> Build it with "RelWithDebInfo", then try it again.

ONNX with RelWithDebInfo mode will crash as well.

**Comment by snnn** [comment]
Then you have the call stack. You will know what happened.

**Comment by jcwchen** [comment]
The crashed stack ([full_stack_log.txt](https://github.com/onnx/onnx/files/6964539/full_stack_log.txt)) is as follows:
```
[0x0]   onnx_cpp2py_export_cp37_win_amd64!__ExceptionPtrRethrow + 0xee   
[0x1]   onnx_cpp2py_export_cp37_win_amd64!std::rethrow_exception + 0xe   
[0x2]   onnx_cpp2py_export_cp37_win_amd64!<lambda_8d89432ffe16d01a50591ffd09e49e26>::<lambda_invoker_cdecl> + 0x47   
[0x3]   torch_python!c10d::GlooDeviceRegistry + 0xe7415   
[0x4]   vcruntime140_1 + 0x1080   
[0x5]   vcruntime140_1!_NLG_Return2 + 0x1555   
[0x6]   ntdll!RtlCaptureContext2 + 0x4a6   
[0x7]   torch_python!torch::FunctionSignature::operator= + 0x98b8e   
...
```
But I still don't know what's going on... It seems like a symbol conflict

**Comment by snnn** [comment]
@garymm do you know why pytorch would call into onnx? torch_python.pyd and onnx_cpp2py_export_cp37_win_amd64.pyd are supposed to be independent.

**Comment by snnn** [comment]
@jcwchen  did you let onnx static link to VC Runtime? What if you change that  to dynamic link?

**Comment by snnn** [comment]
@jcwchen And please copy the onnx_cpp2py_export_cp37_win_amd64.pdb to the installation dir of your onnx, put it in the same folder with onnx_cpp2py_export_cp37_win_amd64.py. Your previous crashed stack doesn't seem to have debug symbols. Let's get a new one.

**Comment by jcwchen** [comment]
> @jcwchen did you let onnx static link to VC Runtime? What if you change that to dynamic link?

Yes I always use USE_MSVC_STATIC_RUNTIME=1 while using static protobuf. If it changes to USE_MSVC_STATIC_RUNTIME=0, the build will fail with static protobuf. The error log is here: https://github.com/jcwchen/onnx/runs/3296316917 (under Build ONNX wheel)

> @jcwchen And please copy the onnx_cpp2py_export_cp37_win_amd64.pdb to the installation dir of your onnx, put it in the same folder with onnx_cpp2py_export_cp37_win_amd64.py. Your previous crashed stack doesn't seem to have debug symbols. Let's get a new one.

Thanks for the reminder. Actually I have already applied .pdb (from RelWithDebInfo mode); otherwise the log will be as follows:
```
[0x0]   onnx_cpp2py_export_cp37_win_amd64!PyInit_onnx_cpp2py_export + 0x25f6ac   
[0x1]   onnx_cpp2py_export_cp37_win_amd64!PyInit_onnx_cpp2py_export + 0x59ab0   
[0x2]   onnx_cpp2py_export_cp37_win_amd64!PyInit_onnx_cpp2py_export + 0x17f09   
[0x3]   torch_python!c10d::GlooDeviceRegistry + 0xe7415   
[0x4]   vcruntime140_1!_CallSettingFrame_LookupContinuationIndex + 0x20   
[0x5]   vcruntime140_1!__FrameHandler4::CxxCallCatchBlock + 0x115   
[0x6]   ntdll!RcFrameConsolidation + 0x6   
[0x7]   torch_python!torch::FunctionSignature::operator= + 0x98b8e   
...
```

**Comment by snnn** [comment]
>  If it changes to USE_MSVC_STATIC_RUNTIME=0, the build will fail with static protobuf

That was because at line https://github.com/jcwchen/onnx/blob/b1c8cdfc137c43364a37a16d9e9b4f3895a2e61d/.github/workflows/release_win.yml#L77 

You set protobuf_MSVC_STATIC_RUNTIME to ON. You need set it to off.

**Comment by snnn** [comment]
Then let's make a PR and fix this. Meantime I will ask VC++ team for help.

**Comment by garymm** [comment]
Apparently PyTorch patches protobuf to make it work in their build:
https://github.com/pytorch/pytorch/blob/master/cmake/ProtoBufPatch.cmake

Perhaps we should do the same?

**Comment by snnn** [comment]
The crash was because there was an C++ exception across DLL boundaries. Either ONNX threw an exception to pytorch, or pytorch threw an exception to ONNX. But these two are independent python packages, they should talk in python, not at such a low level. Recently I'm working on some other things, I didn't have time to debug the issue with @jcwchen. 

I don't like PyTorch's approach. They have the problem only when they dynamical link to protobuf.  Here I want to share some my experience. 

Before starting working on ONNX Runtime,  I worked on Tensorflow Windows support. Someone in our team ported TF to Windows. Tensorflow uses bazel, but bazel wasn't available on Windows yet. So he rewrote all the build scripts in cmake. It worked, with a few limitations. First,  A DLL can only export at most 65536 symbols. TF hits it. So he must use a script to carefully filter symbols out.  He removed the things that may be not needed for building a custom op.  But this is a dirty hack. It is not sustainable when new symbols keep adding in.  And custom authors often see strange link errors. For example, they saw a protobuf symbol was not found, so they add protobuf.lib into the link command, then the build passes. But it will result bigger problem: runtime crash when you pass a protobuf object across DLLs.  If pytorch use the same approach, they will hit the same problem, it is just about when. 

I think, it's a ok if you have to dynamically link to protobuf, but don't go too far away. Don't play the tricks. We should avoid tweaking the symbol names, which might temporarily serve your needs but eventually it will kill us all.

**Comment by jcwchen** [comment]
After investigation, it is likely the reason: https://github.com/pybind/pybind11/issues/1618, https://github.com/pybind/pybind11/issues/2898. Building 2 python modules by pybind11 with static runtime will possibly bump into crash issues while throwing exceptions. Therefore, this PR ( https://github.com/onnx/onnx/pull/3644) should be the proper solution. In that past, ONNX used static MSVC to build python package.

Besides, I have provided @BowenBao the ONNX 1.10 python package built with dynamic MSVC, and he has confirmed that it can resolve the issue he bumped into in this thread and also another crash issue (https://github.com/pytorch/pytorch/issues/62090).

Since it's a fundamental build change and there is a workaround for it (change import order), we won't consume that PR promptly. Let's wait for a while and see whether any user has objection against the PR. If everything goes well, ONNX should consume it before 1.11 release. Thank you all for debugging this issue!
